### PR TITLE
Do not monkeypatch LoginView methods.

### PR DESCRIPTION
**Pull request by D3X**
Related to #273 and freedombox/Plinth#1138

Up for a discussion. If we decide to go with this, I'll add tests to cover both the legacy setting and the new view.

### CODE REVIEW

File: axes/apps.py [hunk 1]
```diff
@@ -1,16 +1,28 @@
...
 class AppConfig(apps.AppConfig):
     name = 'axes'
 
     def ready(self):
-        from django.contrib.auth.views import LoginView
-        from django.utils.decorators import method_decorator
+        from . import signals    # NOQA: Load the signals module to bind the receivers
+
+        if settings.USE_LEGACY_LOGIN_VIEW_PATCH:
+            log.warning('Automatic patching of LoginView using AXES_USE_LEGACY_LOGIN_VIEW_PATCH is deprecated. '
```
Comment (aleksihakli) [inline axes/apps.py]: You should use DeprecationWarning for this.

https://docs.python.org/3.6/library/exceptions.html#DeprecationWarning

**Comment by aleksihakli** [comment]
Looks good aside from using DeprecationWarnings for deprecating features which can be easily tested and are catchable with the `python` `-W` flag. I think implementing pull requests and fixing & testing the infinite redirect bug are essential to this PR, so those should be included in the initial work.

**Comment by camilonova** [comment]
@D3X Could you please take a look again?

**Comment by D3X** [comment]
@camilonova Sure. I kinda forgot about this PR. Will take a look over the weekend.

File: axes/apps.py [hunk 2]
```diff
@@ -1,16 +1,36 @@
...
+from .conf import settings
+
+
+log = logging.getLogger(settings.AXES_LOGGER)
+
+
+def patch_login_view():
+    warnings.warn(
+        'Automatic patching of LoginView using AXES_USE_LEGACY_LOGIN_VIEW_PATCH is deprecated. '
+        'use axes.views.AxesLoginView instead, or add axes.views.AxesLoginMixin to your custom login view.',
```
Comment (aleksihakli) [inline axes/apps.py]: Capitalization `use` -> `Use`.

File: axes/apps.py [hunk 3]
```diff
@@ -1,16 +1,36 @@
...
+        DeprecationWarning
+    )
+    from django.contrib.auth.views import LoginView
```
Comment (aleksihakli) [inline axes/apps.py]: These could be imported on the top of the module, no need to separate the imports module and function level here imo.

File: axes/views.py [hunk 4]
```diff
@@ -0,0 +1,24 @@
+from __future__ import absolute_import
+
+from django.contrib.auth.views import LoginView
+
+from axes.attempts import is_already_locked
```
Comment (aleksihakli) [inline axes/views.py]: This should preferably be `from .attempts import is_already_locked`.

**Comment by D3X** [comment]
@aleksihakli Yeah, I noticed the tests failing. I tried fixing them yesterday, but realised it requires changing a bit more than I initially thought.

The `AdminSite.login` view uses `LoginView` directly, so we'd have to somehow make it use the new `AxesLoginView`. To avoid patching it, we probably have to provide an `AxesAdminSite` class. I don't know how to make it work with other 3rd party apps that register their models using `admin.site.register`, where `admin.site` is the `AdminSite` instance provided by Django.

**Comment by aleksihakli** [comment]
Huh, [looks like](https://github.com/django/django/blob/master/django/contrib/admin/sites.py#L374) [so](https://github.com/django/django/blob/master/django/contrib/admin/sites.py#L349), and the loading of the login and logout views can not be parametrized either via Django other than by subclassing the admin site.

I think the best way to approach this is to both

1. keep the current behaviour and monkey patching of the login views, and
2. offer the new mixins and a custom admin site with documentation.

The first option is what most people will probably use and the second offers more flexibility and customization options. 

Some users will probably benefit from the custom views and admins greatly, and most users will never use them.

The problem with implementing a custom admin is that it increases the amount of code and complexity of this package. Do we want and need that? Should we just document the approaches on implementing the custom views with our mixins and views so people can integrate `axes` into their 3rd party packages and custom admins and implement custom authentication themselves?

In my opinion the custom admin site and mixins are neat and relate closely to the core idea of `axes`, but the growing complexity is a big minus. 

I suggest we should fix the tests and implement the mixins and `AxesLoginView` here but leave the admin site customization up for another PR once this work is merged into upstream.

File: axes/test_settings.py [hunk 5]
```diff
@@ -75,3 +75,4 @@
 LOGIN_REDIRECT_URL = '/admin/'
 
 AXES_FAILURE_LIMIT = 10
+AXES_USE_LEGACY_LOGIN_VIEW_PATCH = False
```
Comment (aleksihakli) [inline axes/test_settings.py]: This should be `True` to allow old test logic to pass.

File: axes/tests/test_apps.py [hunk 6]
```diff
@@ -0,0 +1,25 @@
+import warnings
+
+from django.contrib.auth.views import LoginView
+from django.test import TestCase
+
+from axes.apps import patch_login_view
+
+
+class PatchLoginViewTest(TestCase):
```
Comment (aleksihakli) [inline axes/tests/test_apps.py]: Use [override_settings()](https://docs.djangoproject.com/en/2.0/topics/testing/tools/#django.test.override_settings) to use the newer login view logic.

File: axes/tests/test_views.py [hunk 7]
```diff
@@ -0,0 +1,47 @@
+from unittest.mock import Mock, patch
+
+from django.contrib.auth.models import User
+from django.test import TestCase, RequestFactory
+from django.utils.six.moves import http_client
+
+from axes.views import AxesLoginView
+
+
+class AxesLoginViewTest(TestCase):
```
Comment (aleksihakli) [inline axes/tests/test_views.py]: Use [override_settings()](https://docs.djangoproject.com/en/2.0/topics/testing/tools/#django.test.override_settings) to use the newer login view logic.

File: axes/conf.py [hunk 8]
```diff
@@ -55,3 +55,5 @@ class MyAppConf(AppConf):
     IP_WHITELIST = None
 
     IP_BLACKLIST = None
+
+    USE_LEGACY_LOGIN_VIEW_PATCH = True
```
Comment (aleksihakli) [inline axes/conf.py]: This should probably be just `PATCH_LOGIN_VIEW` which would translate to `AXES_PATCH_LOGIN_VIEW` if we are unsure about the deprecation at the moment.

**Comment by markddavidoff** [comment]
@D3X @aleksihakli maybe take a look at my approach using an authentication backend [here](https://github.com/jazzband/django-axes/pull/315)?

**Pull request closed without merging**